### DISCUSSION TITLE
Assistance with the installation process (101)

### DISCUSSION THREAD

**Discussion by rizzi29** [donmccurdy/glTF-Transform#766] [Q&A] [answered]
Hello and Happy New Year!

I am brand new to this environment and eager to try glTF-Transform.  My apologies in advance for the basic inquiry as I attempt to understand the steps required to download and install the program.  I have tried to install the program utilizing both Windows PowerShell and Command Prompt.  I'm sure that I am not copying and pasting the installation code in the proper format or using the wrong platforms entirely. After 24 hours, of attempts via github (gh repo clone donmccurdy/glTF-Transform) and npm (npm install --global @gltf-transform/cli) , I really could use your assistance.   
Thanks so much and please let me know if I can provide additional details. 

Sincerely,

Michael

**Comment by donmccurdy** [comment]
Hi @rizzi29 — I think the question you are asking is essentially "how can I install this npm package on a Windows PC". It should work the same as any other npm package CLI, but I do not own a Windows PC to guide you on that more specifically. I can say that there should be no need to clone the GitHub repository, just this command should work, if node.js and npm are properly configured in your terminal environment:

```bash
npm install --global @gltf-transform/cli
```

If you're seeing any error messages while trying to install, those are always important to share and/or search for answers related to. :)

**Comment by rizzi29** [comment]
Hi @donmccurdy,  I really appreciate this!  I am now using my Mac :).  I received the following errors after I updated to the latest node.js via: https://nodejs.org/. I downloaded (18.12.1 LTS. 19.3.0 Current and received the same results after: 
npm install --global @gltf-transform/cli

As you mentioned node.js and npm need to be properly configured in my terminal environment and I am guessing that this may be my issue.  Sorry to be a pain...any advice you can provide will be greatly appreciated.

<img width="579" alt="terminal" src="https://user-images.githubusercontent.com/121806176/210465092-64565f8a-b66b-45e1-aed8-001eebe2bd56.png">

**Reply by donmccurdy** [reply, marked as answer]
Looks like you're close! The error you're seeing indicates that, whatever you might have installed, Node.js v12 is what's running in your terminal. This project requires Node.js v14 or newer. There's more than one way to approach this, but my recommendation would be to install nvm ("Node Version Manager") so you have direct control of which version gets used:

https://github.com/nvm-sh/nvm

Then you can do something like:

```shell
nvm install 18
nvm alias default 18
npm install --global @gltf-transform/cli
```

**Reply by rizzi29** [reply]
Hi @donmccurdy,
Thanks so much for walking me through this!  Means the world. Looks like a bit more progress for sure.  Just a few warnings that I can could really use your advice in order to solve.  Thanks again for sticking with me... :)

<img width="587" alt="Terminal 2" src="https://user-images.githubusercontent.com/121806176/210583027-3d68bb48-e221-459c-a12a-1d2607ea34ad.png">

**Reply by donmccurdy** [reply]
I think that's just a warning, you might be able to use gltf-transform already! Try `gltf-transform --help` and see what comes out? The warning is about the squoosh library not working in Node.js v18, and squoosh is needed only if you're trying to do image compression (JPEG, PNG, WebP) with this library. That'll be fixed in an upcoming release, or some workarounds are possible in the meantime if you need image compression sooner.

**Reply by rizzi29** [reply]
Thank so much! I think that this is almost working.  My main objective is to reduce the overall size of an .obj file (932MB).  Many thanks for your help and patience... :)
<img width="568" alt="Terminal 4" src="https://user-images.githubusercontent.com/121806176/210865880-5aa0f922-34c2-490b-905c-504fd32b1ff5.png">

**Reply by donmccurdy** [reply]
I think the errors and warnings there can be ignored, though switching to Node.js v16 would resolve them if you would like.

Try `gltf-transform inspect /Users/success/Desktop/Rudolf/Rudolpha.glb --format md`, what output do you get? If `resize` has no effect then it's likely that either (a) the model has no textures, or (b) the textures are already under 1024x1024. In either case the 'inspect' output should help guess what to try next.

**Reply by rizzi29** [reply]
Ok, I believe we have something.  :) I switched to Node.js v16.  I was under the assumption that the program would reduce both meshes and textures.  Based on the this command, meshes are definitely the culprit, here.  In the case when I do have a successful conversion utilizing a different file, would the program display the path (or folder name) for the output?  Just not sure where to find the output file.  Thank you @donmccurdy!

<img width="939" alt="Terminal 5" src="https://user-images.githubusercontent.com/121806176/211035389-82143b70-44ff-4547-a729-75da74c6e416.png">

**Reply by donmccurdy** [reply]
The `inspect` command does not output a file, but in other commands you specify the output path yourself with something like `gltf-transform <command> <input> <output>`. It does appear that the meshes are the problem here! 

glTF Transform offers two basic ways to reduce the amount of mesh data:

1. compress the existing vertices
2. reduce the number of vertices

Currently this tool is better at (1) than (2). It offers more than one type of compression, but the simplest to try would be:

```bash
gltf-transform draco path/to/input.glb path/to/output.glb
```

This applies [Draco compression](https://github.com/google/draco) to the mesh geometry.

However — I should caution that 25M vertices is a lot, regardless of the size of the file! Many viewers might have trouble rendering this quickly, and compressing the data does not help that problem, because the vertices still have to be decompressed before rendering. To reduce the number of vertices, if that's a concern, you need to _simplify_ or _decimate_ the mesh, before compressing it:

```bash
gltf-transform weld path/to/input.glb path/to/output.glb
gltf-transform simplify path/to/output.glb path/to/output.glb --error 0.0001 --ratio 0.01 --lock-border
gltf-transform draco path/to/output.glb path/to/output.glb
```

Sometimes simplification can create visible errors in the mesh, though. If you run into that, you might need to consider a more full-featured simplification tool like Blender's "Decimate" modifier, or Rapid Compact.

**Reply by rizzi29** [reply]
Thank you so much for this!  I really appreciate your time and patience.  I'll keep working on this and start with lower res 3D scans. :) Thank you @donmccurdy!
